lib/Optimizer/Transforms/DependencyAnalysis.cpp

```diff
@@ -0,0 +1,2561 @@
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "PassDetails.h"
+#include "cudaq/Frontend/nvqpp/AttributeNames.h"
+#include "cudaq/Optimizer/Dialect/CC/CCOps.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeDialect.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeTypes.h"
+#include "cudaq/Optimizer/Transforms/Passes.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/Passes.h"
+
+#define DEBUG_TYPE "dep-analysis"
+
+using namespace mlir;
+
+#define RAW(X) quake::X
+#define RAW_MEASURE_OPS MEASURE_OPS(RAW)
+#define RAW_GATE_OPS GATE_OPS(RAW)
+#define RAW_QUANTUM_OPS QUANTUM_OPS(RAW)
+
+//===----------------------------------------------------------------------===//
+// Generated logic
+//===----------------------------------------------------------------------===//
+namespace cudaq::opt {
+#define GEN_PASS_DEF_DEPENDENCYANALYSIS
+#include "cudaq/Optimizer/Transforms/Passes.h.inc"
+} // namespace cudaq::opt
+
+namespace {
+typedef size_t PhysicalQID;
+typedef size_t VirtualQID;
+
+size_t getOperandIDXFromResultIDX(size_t resultidx, Operation *op) {
+  if (isa<RAW_MEASURE_OPS>(op))
+    return 0;
+  if (isa<quake::SwapOp>(op))
+    return (resultidx == 0 ? 1 : 0);
+  // Currently, all classical operands precede all quantum operands
+  for (auto type : op->getOperandTypes()) {
+    if (!quake::isQuantumType(type))
+      resultidx++;
+    else
+      break;
+  }
+  return resultidx;
+}
+
+size_t getResultIDXFromOperandIDX(size_t operand_idx, Operation *op) {
+  if (isa<RAW_MEASURE_OPS>(op))
+    return 1;
+  if (isa<quake::SwapOp>(op))
+    return (operand_idx == 0 ? 1 : 0);
+  // Currently, all classical operands precede all quantum operands
+  for (auto type : op->getOperandTypes()) {
+    if (!quake::isQuantumType(type))
+      operand_idx--;
+    else
+      break;
+  }
+  return operand_idx;
+}
+
+class LifeTime {
+protected:
+  uint begin;
+  uint end;
+
+public:
+  LifeTime(uint begin, uint end) : begin(begin), end(end) {
+    assert(end >= begin && "invalid lifetime");
+  };
+
+  bool isAfter(LifeTime *other) { return begin > other->end; }
+
+  bool isOverlapping(LifeTime *other) {
+    return !isAfter(other) && !other->isAfter(this);
+  }
+
+  uint distance(LifeTime *other) {
+    if (isOverlapping(other))
+      return 0;
+    return std::max(begin, other->begin) - std::min(end, other->end);
+  }
+
+  void combine(LifeTime *other) {
+    begin = std::min(begin, other->begin);
+    end = std::max(end, other->end);
+  }
+
+  uint getBegin() { return begin; }
+  uint getEnd() { return end; }
+};
+
+class LifeTimeAnalysis {
+private:
+  StringRef name;
+  SmallVector<LifeTime *> lifetimes;
```

Comment:
curious: why is this a vector of raw pointers? what if I use a `SmallVector<LifeTime>` instread?

Reply:
Yes, @bmhowe23 and I had talked about doing that instead, which will prevent some potential memory leaks. I'll work on that after I address the documentation.

Reply:
Fixed